Adds a "pred" macro for better filter definition

sritchie opened the pull request:
The code looked pretty good. I added two main items:

### Column Protocol 

The "column" protocol that allows you to define a column based on some comparison value and a column name. The implementation dispatches on the type of the comparison value. For example:

```clojure
parquet-thrift-cascalog.filter> (column (byte-array [1 2 3]) "bytes")
#<BinaryColumn column(bytes)>
parquet-thrift-cascalog.filter> (column 10 "face")
#<LongColumn column(face)>
```

It's also defined for a value that's already a column. That implementation just clones the column type but changes the column name:

```clojure
parquet-thrift-cascalog.filter> (column (column 10 "face") "cake")
#<LongColumn column(cake)>
```

I DON'T have a String implementation defined, since Parquet doesn't give us a String column yet.

### "Pred" macro

Not a Cascalog predicate macro :) Just wanted to make things confusing.

The `pred` macro inside of the `filter` namespace allows you to use Clojure's `<=`, `<` etc comparators instead of `ltEq`, `lt` etc. The macro replaces the Clojure forms with the parquet forms, allowing you to declare filters like this:

```clojure
parquet-thrift-cascalog.filter> 
(pred
 (or (> "cake" 100.234)
     (and (not (< "face" 10))
          (>= "face" 13))))
#<Or or(gt(cake, 100.234), and(not(lt(face, 10)), gteq(face, 13)))>
```

I added a test namespace for this too. I can add a few more tests if you're into this.

alexrobbins commented:
The changes look good to me. I'm going to copy your pred macro documentation into the readme.

Also, could we make the string stuff work transparently by having the macro automatically wrap string comparisons in the string->binary call and then use a binarycolumn? I think that is going to be a common use case for us.

sritchie commented:
Yeah, I think we definitely could. I'm just not sure how the string comparisons work when you actually run your job, so I didn't know if setting the column was enough. I guess strings are always stored as byte arrays, so that'd be great. Want me to make that change?

alexrobbins commented:
I looked around a little yesterday and it looks like sorting the bytes behind a UTF-8 string has the same output as sorting the strings lexicographically.

http://www.icu-project.org/docs/papers/utf16_code_point_order.html
https://en.wikipedia.org/wiki/UTF-8#Advantages

So we should be able to coerce the string to binary and then use the comparisons naively.

If you wanted to add that, that'd be great.

sritchie commented:
Done, added the string implementation. I also fixed what looked like a bug with my binary comparison. The "binaryColumn" actually takes an instance of Binary, so I added conversions to that type from the byte array, from `String` and from `ByteBuffer`.

Go ahead and merge if you're into it. This should be helpful!

alexrobbins commented:
Now that we have a macro wrapper, I'm starting to feel like the convenience functions I added aren't that useful. What do you think about the macro wrapper just expanding into calls directly into the parquet FilterApi class? Same with your protocol work. I can make the change if you think it makes sense.

sritchie commented:
Up to you! It's not a big deal to have the duplication, I think, and it might be convenient to have the wrappers around if anyone else wants to play with a different DSL.

One advantage the wrapper has is that precondition. Right now you have to supply the column name first, so `(< 10 "col-name")` will throw an assertion error. Until we beef up the macro to rewrite backwards conditions I think we might want to keep the wrappers.

alexrobbins commented:
Sounds good.

Backwards conditions will be tough because they'd be ambiguous for string columns unless the column name or value had already been converted.

alexrobbins commented:
Released as 0.1.3. I had to make some changes because the value coercion happening in the column protocol functions was being discarded. I set up a `ParquetValue` protocol that converts input into a parquet recognized value. It does the conversion work that `Column` was trying to do.

I also changed the defpred produced functions to accept a string name or a column. This allows predicates to work with nil values, which don't provide type information, using a form like:

`(pred (= (string-column "name") nil)`

Anyway, thanks for the work on this. I think it is a huge improvement to the filter interface. Type inference FTW.

sritchie commented:
Nice! Fun working with you on this one. I'll add a link to the Cascalog docs too, if that's alright with you.

—
Sent from Mailbox

On Thu, Feb 26, 2015 at 9:39 AM, Alex Robbins notifications@github.com
wrote:

> Released as 0.1.3. I had to make some changes because the value coercion happening in the column protocol functions was being discarded. I set up a `ParquetValue` protocol that converts input into a parquet recognized value. It does the conversion work that `Column` was trying to do.
> I also changed the defpred produced functions to accept a string name or a column. This allows predicates to work with nil values, which don't provide type information, using a form like:
> `(pred (= (string-column "name") nil)`
> 
> ## Anyway, thanks for the work on this. I think it is a huge improvement to the filter interface. Type inference FTW.
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/alexrobbins/parquet-thrift-cascalog/pull/3#issuecomment-76212376

alexrobbins commented:
Sure, sounds good.
